File: 146. LRU Cache.js
/**
 * @constructor
 */
var LRUCache = function(capacity) {
    this.size = capacity;
    this.dic = {};//map
    this.head = new DBListNode(-1);
    this.trail = new DBListNode(-1);
    //-------------------//
    this.head.next = this.trail;
    this.trail.prev = this.head;
};

/**
 * @param {number} key
 * @returns {number}
 */
LRUCache.prototype.get = function(key) {
    if(!(key in this.dic)) return -1;
    var node = this.dic[key];
    var value = node.val;
    //-------------------//
    var prev = node.prev;
    var next = node.next;
    prev.next = next;
    next.prev = prev;
    //------------------//
    prev = this.head;
    next = head.next;
    prev.next = node;
    node.next = next;
    next.prev = node;
    node.prev = head;
    return value;
};

/**
 * @param {number} key
 * @param {number} value
 * @returns {void}
 */
LRUCache.prototype.set = function(key, value) {
    if(key in this.dic) {
        this.dic[key] = value;
        this.update(key);
    }
    else {
        if(Object.keys(this.dic).length == this.size) {
            var node = this.dic[this.trail.prev];
            this.dic.delete(node.key);
            node.key = key;
            node.val = value;
            this.dic[key] = node;
            update(key);
        }
        else {
            var tmp = new DBListNode(key, value);
            this.dic[key] = tmp;
            var first = this.head.next;
            this.head.next = tmp;
            tmp.next = first;
            first.prev = tmp;
            tmp.prev = this.head;
        }
    }
}

var DBListNode = function(val, key) {
    this.val = val;
    this.key = key;
    this.prev = null;
    this.next = null;
}

File: Leetcode/Java/146. LRU Cache.java
public class LRUCache {
    int size;
    Map<Integer, ValueNode> map;
    ValueNode start;
    ValueNode end;
    public LRUCache(int capacity) {
        size = capacity;
        map = new HashMap<Integer, ValueNode>();
        start = new ValueNode(-1,-1);
        end = new ValueNode(-1,-1);
        start.next = end;
        end.prev = start;
    }

    public int get(int key) {
        if(!map.containsKey(key)) return -1;
        ValueNode value = map.get(key);
        ValueNode prev = value.prev;
        ValueNode next = value.next;
        prev.next = next;
        next.prev = prev;
        setFirst(value);
        return value.val;
    }

    public void set(int key, int value) {
        if(map.containsKey(key)) {
            map.get(key).val = value;
            get(key);
        }
        else {
            if(map.size() == size) removeLast();
            ValueNode node = new ValueNode(value, key);
            setFirst(node);
            map.put(key, node);
        }
    }

    private void setFirst(ValueNode node) {
        ValueNode first = start.next;
        start.next = node;
        node.prev = start;
        node.next = first;
        first.prev = node;
    }

    private void removeLast() {
        ValueNode last = end.prev;
        int key = last.key;
        end = last;
        end.next = null;
        map.remove(key);
    }
    class ValueNode {
        int key;
        int val;
        ValueNode prev;
        ValueNode next;
        public ValueNode(int val, int key) {
            this.val = val;
            this.key = key;
            prev = null;
            next = null;
        }
    }
}
